File: videoprokat/Catalog.cpp
#include "Catalog.hpp"
#include "VideoCarrier.hpp"

#include <algorithm>

void Catalog::addItem(VideoCarrier* item)
{
    if (item) m_items.push_back(item);
}

VideoCarrier* Catalog::findItemByNumber(int number) const
{
    for (auto* item : m_items) {
        if (item && item->getInventoryNumber() == number) return item;
    }
    return nullptr;
}

std::vector<VideoCarrier*> Catalog::findItemsByTitle(const std::string& title) const
{
    std::vector<VideoCarrier*> result;
    for (auto* item : m_items) {
        if (item && item->getTitle() == title) result.push_back(item);
    }
    return result;
}

std::vector<VideoCarrier*> Catalog::getAvailableItems() const
{
    std::vector<VideoCarrier*> result;
    for (auto* item : m_items) {
        if (item && item->isAvailable()) result.push_back(item);
    }
    return result;
}

File: Catalog.java
package videoprokat;

import java.util.List;

/**
 * Абстрактный класс, описывающий каталог видеоносителей.
 * Содержит базовую структуру и методы для работы с каталогом.
 */
public abstract class Catalog {

    protected List<VideoCarrier> items;

    /**
     * Конструктор по умолчанию.
     */
    protected Catalog() {
        // Инициализация коллекции выполняется в реализации
    }

    /**
     * Добавить носитель в каталог.
     */
    public abstract void addItem(VideoCarrier item);

    /**
     * Найти носитель по инвентарному номеру.
     */
    public abstract VideoCarrier findItemByNumber(int number);

    /**
     * Найти все носители по названию.
     */
    public abstract List<VideoCarrier> findItemsByTitle(String title);

    /**
     * Получить список доступных носителей.
     */
    public abstract List<VideoCarrier> getAvailableItems();
}